src/libraries/System.Collections/src/System/Collections/Generic/PriorityQueue.cs

```diff
@@ -0,0 +1,514 @@
+        public PriorityQueue(int initialCapacity, IComparer<TPriority>? comparer)
+        {
+            if (initialCapacity < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(initialCapacity));
+            }
+
+            this.nodes = new List<(TElement, TPriority)>(initialCapacity);
+            this.UnorderedItems = new UnorderedItemsCollection(this.nodes);
+            this.Comparer = comparer ?? Comparer<TPriority>.Default;
+        }
+
+        /// <summary>
+        /// Creates a priority queue populated with the specified elements and priorities.
+        /// </summary>
+        public PriorityQueue(IEnumerable<(TElement element, TPriority priority)> items)
+        {
+            if (items is null)
+            {
+                throw new ArgumentNullException(nameof(items));
+            }
+
+            this.nodes = new List<(TElement, TPriority)>(items);
+            this.UnorderedItems = new UnorderedItemsCollection(this.nodes);
+            this.Comparer = Comparer<TPriority>.Default;
+
+            if (this.nodes.Count > 1)
+            {
+                this.Heapify();
+            }
+        }
+
+        /// <summary>
+        /// Creates a priority queue populated with the specified elements and priorities,
+        /// and with the specified priority comparer.
+        /// </summary>
+        public PriorityQueue(IEnumerable<(TElement element, TPriority priority)> items, IComparer<TPriority>? comparer)
+        {
+            if (items is null)
+            {
+                throw new ArgumentNullException(nameof(items));
+            }
+
+            this.nodes = new List<(TElement, TPriority)>(items);
+            this.UnorderedItems = new UnorderedItemsCollection(this.nodes);
+            this.Comparer = comparer ?? Comparer<TPriority>.Default;
+
+            if (this.nodes.Count > 1)
+            {
+                this.Heapify();
+            }
+        }
+
+        /// <summary>
+        /// Gets the current amount of items in the priority queue.
+        /// </summary>
+        public int Count => this.nodes.Count;
+
+        /// <summary>
+        /// Gets the priority comparer of the priority queue.
+        /// </summary>
+        public IComparer<TPriority> Comparer { get; }
+
+        /// <summary>
+        /// Enqueues the specified element and associates it with the specified priority.
+        /// </summary>
+        public void Enqueue(TElement element, TPriority priority)
+        {
+            if (element is null)
+            {
+                throw new ArgumentNullException(nameof(element));
+            }
+
+            if (priority is null)
+            {
+                throw new ArgumentNullException(nameof(priority));
+            }
+
+            // Add the node at the end
+            var node = (element, priority);
+            this.nodes.Add(node);
+
+            // Restore the heap order
+            var lastNodeIndex = this.GetLastNodeIndex();
```

Comment:
We avoid using `var` unless the type is obvious from the RHS, namely with a `new` or explicit cast.